mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala

```diff
@@ -449,6 +449,16 @@ class SparseVector(
   override def toString: String =
     "(%s,%s,%s)".format(size, indices.mkString("[", ",", "]"), values.mkString("[", ",", "]"))
 
```

Comment:
Yeah I was thinking of performance too. That is a fair point of course about not changing the semantics. I think the `hashCode` impl could be changed in both places to produce the same result while being faster for sparse vectors: what if the hash were defined only over nonzero entries? I think `equals` could likewise be sped up also for the case of comparing sparse to dense? that is I think this can be taken a step further than just optimizing sparse == sparse.